src/server/kernel/wseat.cpp

```diff
@@ -621,6 +726,105 @@ void WSeat::notifyFrame(WCursor *cursor)
     d->doNotifyFrame();
 }
 
+// top level touch event
+
+void WSeat::notifyTouchDown(WCursor *cursor, WInputDevice *device, int32_t touch_id, uint32_t time_msec)
+{
+    W_D(WSeat);
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    Q_ASSERT(qwDevice);
+    const QPointF &globalPos = cursor->position();
+
+    qDebug() << "TouchDown: tid" <<  touch_id << " pos:" << globalPos << " time:" << time_msec;
+    auto *state = &d->devState[qwDevice];
+    QWindowSystemInterface::TouchPoint *tp = state->point(touch_id);
+    if (tp) {
+        qWarning("Incorrect touch state");
+    } else {
+        QWindowSystemInterface::TouchPoint newTp;
+        newTp.id = touch_id;
+        newTp.state = QEventPoint::Pressed;
+        newTp.area = QRect(0, 0, 8, 8);
+        newTp.area.moveCenter(globalPos);
+        state->m_points.append(newTp);
+    }
+}
+
+void WSeat::notifyTouchMotion(WCursor *cursor, WInputDevice *device, int32_t touch_id, uint32_t time_msec) {
+
+    W_DC(WSeat);
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    Q_ASSERT(qwDevice);
+
+    const QPointF &globalPos = cursor->position();
+qDebug() << "TouchMotion: tid" <<  touch_id << " pos:" << globalPos << " time:" << time_msec;
+
+    auto *state = &d->devState[qwDevice];
+    QWindowSystemInterface::TouchPoint *tp = state->point(touch_id);
+    if (tp) {
+        auto tmpState = QEventPoint::Updated;
+        if (tp->area.center() == globalPos)
+            tmpState = QEventPoint::Stationary;
+        else
+            tp->area.moveCenter(globalPos);
+        // 'down' may be followed by 'motion' within the same "frame".
+        // Handle this by compressing and keeping the Pressed state until the 'frame'.
+        if (tp->state != QEventPoint::Pressed && tp->state != QEventPoint::Released)
+            tp->state = tmpState;
+
+        qDebug() << "MoveUpdate2: " << tp->state << tmpState;
+    } else {
+        qWarning("Inconsistent touch state (got 'motion' without 'down')");
+    }
+}
+
+void WSeat::notifyTouchUp(WCursor *cursor, WInputDevice *device, int32_t touch_id, uint32_t time_msec)
+{
+    W_DC(WSeat);
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    Q_ASSERT(qwDevice);
+qDebug() << "TouchUp: tid" <<  touch_id  << " time:" << time_msec;
+
+    auto *state = &d->devState[qwDevice];
+    QWindowSystemInterface::TouchPoint *tp = state->point(touch_id);
+    if (tp) {
+        tp->state = QEventPoint::Released;
+        // There may not be a Frame event after the last Up. Work this around.
+        QEventPoint::States s;
+        for (auto poiner : state->m_points) {
+            s |= poiner.state;
+        }
+        if (s == QEventPoint::Released)
+            notifyTouchFrame();
```

Comment:
Why need manual send touch frame? Does it should only from WCursor?

Reply:
There may not be a Frame event after the last Up. Work this around.
https://codebrowser.dev/qt6/qtbase/src/platformsupport/input/libinput/qlibinputtouch.cpp.html#147

Reply:
Add some annotations here